radixdlt/src/main/java/com/radixdlt/consensus/ChainedBFT.java

```diff
@@ -88,47 +91,53 @@ public ChainedBFT(
 			.autoConnect(2);
 
 		final Observable<Event> epochs = epochEvents
-			.map(o -> new Event(EventType.EPOCH, o))
-			.subscribeOn(this.singleThreadScheduler);
+			.map(o -> new Event(EventType.EPOCH, o));
 
-		final Observable<EventCoordinator> epochCoordinators = epochEvents
+		final Observable<EventCoordinator> eventCoordinators = epochEvents
 			.map(epochManager::nextEpoch)
 			.startWithItem(epochManager.start())
 			.doOnNext(EventCoordinator::start)
-			.publish()
-			.autoConnect(2); // timeouts and consensusMessages below
-
-		final Observable<Event> timeouts = Observable.combineLatest(
-			epochCoordinators,
-			this.pacemakerRx.localTimeouts(),
-			(e, timeout) -> {
-				e.processLocalTimeout(timeout);
-				return new Event(EventType.LOCAL_TIMEOUT, timeout);
-			}
-		).subscribeOn(this.singleThreadScheduler);
-
-		final Observable<Event> consensusMessages = Observable.combineLatest(
-			epochCoordinators,
-			this.network.consensusEvents(),
-			(e, msg) -> {
-				final EventType eventType;
-				if (msg instanceof NewView) {
-					e.processNewView((NewView) msg);
-					eventType = EventType.NEW_VIEW_MESSAGE;
-				} else if (msg instanceof Proposal) {
-					e.processProposal((Proposal) msg);
-					eventType = EventType.PROPOSAL_MESSAGE;
-				} else if (msg instanceof Vote) {
-					e.processVote((Vote) msg);
-					eventType = EventType.VOTE_MESSAGE;
-				} else {
-					throw new IllegalStateException("Unknown Consensus Message: " + msg);
-				}
-
-				return new Event(eventType, msg);
-			}
-		).subscribeOn(this.singleThreadScheduler);
-
-		return Observable.merge(epochs, timeouts, consensusMessages);
+			.replay(1)
+			.autoConnect(); // timeouts, rpcs and consensusMessages below
+
+		// Need to ensure that first event coordinator is emitted otherwise we may not process
+		// initial events due to the .withLatestFrom() drops events
+		final Completable firstEventCoordinator = Completable.fromSingle(eventCoordinators.firstOrError());
+		final Observable<Event> ecMessages = firstEventCoordinator
+			.andThen(
+				Observable.merge(
+					this.pacemakerRx.localTimeouts().observeOn(this.singleThreadScheduler),
+					this.network.consensusEvents().observeOn(this.singleThreadScheduler),
+					this.network.rpcRequests().observeOn(this.singleThreadScheduler)
+				)
+					.withLatestFrom(
+						eventCoordinators,
+						(msg, e) -> {
+							final EventType eventType;

```

Comment:
Holy mother of indentation. Maybe we can extract this into a method?

Reply:
^^ yeap.